### ISSUE TITLE
Add support for static website to azurerm_storage_account

### ISSUE THREAD

**Issue by hughtopping** [terraform-providers/terraform-provider-azurerm#1903]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Static website hosting for Azure Storage is now in public preview. [See here for Azure announcement.](https://azure.microsoft.com/en-us/blog/azure-storage-static-web-hosting-public-preview/)

Currently this can be added with the Azure Portal UI, but is missing from the `azurerm_storage_account` Terraform resource. 

### New or Affected Resource(s)

* azurerm_storage_account

### Potential Terraform Configuration

```hcl
resource "azurerm_storage_account" "static_website" {
  account_replication_type  = "RAGRS"
  account_tier              = "Standard"
  account_kind              = "StorageV2"
  location                  = "westeurope"
  name                      = "static_website_storage_account"
  resource_group_name       = "static_website_rg"
  enable_https_traffic_only = true

  static_website {
    enabled             = true
    index_document_name = "index.html"
    error_document_path = "error.html"
  }
}
```

### References

* https://azure.microsoft.com/en-us/blog/azure-storage-static-web-hosting-public-preview/

**Comment by 0x7f** [comment]
For those who can not wait until this feature lands in stable terraform and still want to manage their storage with static website hosting enabled via terraform, I have a possible workaround.

First, I tried creating [azurerm_template_deployment](https://www.terraform.io/docs/providers/azurerm/r/template_deployment.html) and enable the static website feature via an ARM template. Unfortunately this is not possible, because static website feature is also not available via arm templates yet, see [here](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Storage/2018-07-01/storageAccounts#StorageAccountPropertiesCreateParameters). It is only possible via Blob Service REST API since API Version [2018-03-28](https://docs.microsoft.com/en-us/rest/api/storageservices/set-blob-service-properties#request) which you can use via azure cli.

So the workaround is to run enable static website via azure cli using terraform's [`local-exec`](https://www.terraform.io/docs/provisioners/local-exec.html) provisioner. Based on your example it could look like:

```tf
resource "azurerm_storage_account" "static_website" {
  account_replication_type  = "RAGRS"
  account_tier              = "Standard"
  account_kind              = "StorageV2"
  location                  = "westeurope"
  name                      = "static_website_storage_account"
  resource_group_name       = "static_website_rg"
  enable_https_traffic_only = true

  provisioner "local-exec" {
    command = <<EOF
az extension add --name storage-preview
az storage blob service-properties update \
  --subscription "${var.subscription_id}" \
  --account-name static_website_storage_account \
  --static-website \
  --404-document 404.html \
  --index-document index.html
EOF
  }
}
```

This will make sure that the static website feature is enabled via cli right after the storage account resource is created.

Downside of this is that the command will never be executed when you import an existing storage account into your tfstate and just want to enable static website later. To allow this, you can move the `local-exec` into a dedicated resource and let it depend on the storage account resouce, e.g.:

```tf
resource "null_resource" "static_website_cmd" {
  provisioner "local-exec" {
    command = <<EOF
az extension add --name storage-preview
az storage blob service-properties update \
  --subscription "${var.subscription_id}" \
  --account-name static_website_storage_account \
  --static-website \
  --404-document 404.html \
  --index-document index.html
EOF
  }

  depends_on = ["azurerm_storage_account.static_website"]
}
```

This way you are more flexible, e.g. you can taint the null_resource (which manually marks the resource for redeployment in the tfstate) and by that triggers a re-execution of the command. Tainting can be done like this: 

```sh
$ terraform taint -module my_module null_resource.static_website_cmd
$ terraform apply
...
-/+ module.my_module.null_resource.static_website_cmd (tainted) (new resource required)
      id:                               "123456789123456789" => <computed> (forces new resource)
...
```

**Comment by katbyte** [comment]
Hi @hughtopping, @0x7f 

I looked at the [latest version](https://github.com/Azure/azure-sdk-for-go/blob/ca4654c50e30248fa542c1f0dd07fbde21d9b9a2/services/storage/mgmt/2018-07-01/storage/accounts.go) of the SDK & storage API and I don't see any sign of these properties 🙁 Until they are in the SDK this is unfortunately blocked.

**Comment by 0x7f** [comment]
Hi @katbyte true, the Azure Go SDK does not support it, but there is also the [azure-storage-blog-go](https://github.com/Azure/azure-storage-blob-go) SDK that actually has support for it, see [here](https://github.com/Azure/azure-storage-blob-go/blob/5152f14ace1c6db66bd9cb57840703a8358fa7bc/azblob/zz_generated_models.go#L4560-L4568). To be honest, I don't know the difference between the two SDKs, especially since both are provided by Azure. Maybe it is possible to use that one instead?

**Comment by mikaelkrief** [comment]
Here the code provided by Microsoft blog using the provisionner

```
resource "azurerm_storage_account" "webblob" {
  name                     = "${var.dns_name}"
  location                 = "${azurerm_resource_group.demo-rg.location}"
  resource_group_name      = "${azurerm_resource_group.demo-rg.name}"
  account_kind             = "StorageV2"
  account_tier             = "Standard"
  account_replication_type = "LRS"

   provisioner "local-exec" {
    command = "az storage blob service-properties update --account-name ${azurerm_storage_account.webblob.name} --static-website  --index-document index.html --404-document 404.html"
  }
}
```
source : [terraform-jamstack-azure-gatsby-azure-pipelines-git](https://cloudblogs.microsoft.com/opensource/2018/11/16/terraform-jamstack-azure-gatsby-azure-pipelines-git/)

**Comment by kvaes** [comment]
Good suggestion! Just tested it and I found a few issues with it...

- The ["blob service-properties update"](https://docs.microsoft.com/en-us/cli/azure/storage/blob/service-properties?view=azure-cli-latest) does not exist?

- When the provisioner is added to an existing resource, then the command is not executed. As the provisioner is run [after creation](https://www.terraform.io/docs/provisioners/local-exec.html)...

- When the storage account was just deleted, then the following error occurs (so keep some "cooldown" time in mind) ; 

> * azurerm_storage_account.storage: Error creating Azure Storage Account "vmsnoozerdev": storage.AccountsClient#Create: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="StorageAccountNotFound" Message="The storage account vmsnoozerdev was not found."

- The "az login" part is quite key. Either do it (as the blog describes) via a prior build step, or within the command itself. Examply by using a [service principal](https://docs.microsoft.com/en-us/cli/azure/authenticate-azure-cli?view=azure-cli-latest#sign-in-with-a-service-principal) ; 

```
resource "azurerm_storage_account" "storage" {
  name                     = "${var.prefix_workload}${var.prefix_environment}"
  resource_group_name      = "${azurerm_resource_group.storage.name}"
  location                 = "${azurerm_resource_group.storage.location}"
  account_tier             = "${var.storage_tier}"
  account_replication_type = "${var.storage_resiliency}"

  provisioner "local-exec" {
    command = "az login  --service-principal -u \"${var.armclientid}\" -p \"${var.armclientsecret}\" --tenant \"${var.armtenantid}\" | az storage blob service-properties update --account-name ${azurerm_storage_account.storage.name} --static-website  --index-document index.html --404-document 404.html"
  }
}
```

**Comment by pburkholder** [comment]
@pwelch @yashkhemani Here's the Azure issue I was up against this week. If this were a native Terraform feature, instead of a `local-exec` it would make this Microsoft blog, https://cloudblogs.microsoft.com/opensource/2018/11/16/terraform-jamstack-azure-gatsby-azure-pipelines-git/, I lot less janky. cc: @yenisha

**Comment by lenisha** [comment]
@pburkholder, yes waiting for this feature.

**Comment by Djiit** [comment]
So does this mean it's being ignored by hashicorp devs ?

**Comment by tombuildsstuff** [comment]
@Djiit no, unfortunately there's multiple blockers preventing us from adopting the new Storage SDK ([as mentioned above](https://github.com/terraform-providers/terraform-provider-azurerm/issues/1903#issuecomment-443642722)) - whilst we're working on a solution for this, unfortunately we don't have a timeline.

**Comment by jungopro** [comment]
> Good suggestion! Just tested it and I found a few issues with it...
> 
> * The ["blob service-properties update"](https://docs.microsoft.com/en-us/cli/azure/storage/blob/service-properties?view=azure-cli-latest) does not exist?
> * When the provisioner is added to an existing resource, then the command is not executed. As the provisioner is run [after creation](https://www.terraform.io/docs/provisioners/local-exec.html)...
> * When the storage account was just deleted, then the following error occurs (so keep some "cooldown" time in mind) ;
> 
> > * azurerm_storage_account.storage: Error creating Azure Storage Account "vmsnoozerdev": storage.AccountsClient#Create: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="StorageAccountNotFound" Message="The storage account vmsnoozerdev was not found."
> 
> * The "az login" part is quite key. Either do it (as the blog describes) via a prior build step, or within the command itself. Examply by using a [service principal](https://docs.microsoft.com/en-us/cli/azure/authenticate-azure-cli?view=azure-cli-latest#sign-in-with-a-service-principal) ;
> 
> ```
> resource "azurerm_storage_account" "storage" {
>   name                     = "${var.prefix_workload}${var.prefix_environment}"
>   resource_group_name      = "${azurerm_resource_group.storage.name}"
>   location                 = "${azurerm_resource_group.storage.location}"
>   account_tier             = "${var.storage_tier}"
>   account_replication_type = "${var.storage_resiliency}"
> 
>   provisioner "local-exec" {
>     command = "az login  --service-principal -u \"${var.armclientid}\" -p \"${var.armclientsecret}\" --tenant \"${var.armtenantid}\" | az storage blob service-properties update --account-name ${azurerm_storage_account.storage.name} --static-website  --index-document index.html --404-document 404.html"
>   }
> }
> ```

Hello @kvaes and @tombuildsstuff 

I followed up on this and was able to successfully configure the storage account for website hosting.

Were you able to figure out a way to output the URL? 
I know I can use the `az cli` for this as well as documented [here](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blob-static-website-how-to#cli-find-url) but I was wondering if maybe this is already supported natively within the storage account resource in terraform... 

Looking at the [resource documentation](https://www.terraform.io/docs/providers/azurerm/r/storage_account.html) I thought that maybe _primary_blob_endpoint_ was the way to go but this is not the correct endpoint and I couldn't find anything else...

Any ideas?

Thanks
Omer

**Comment by hazzik** [comment]
>Were you able to figure out a way to output the URL?

I'm using "data" "storage_account" 

"primary_web_endpoint" output.

https://www.terraform.io/docs/providers/azurerm/d/storage_account.html#primary_web_endpoint

**Comment by jungopro** [comment]
> > Were you able to figure out a way to output the URL?
> 
> I'm using "data" "storage_account" "primary_web_endpoint" output attibute.
> 
> https://www.terraform.io/docs/providers/azurerm/d/storage_account.html#primary_web_endpoint

missed that. worked like a charm. thanks!

**Comment by StefanSchoof** [comment]
Since it looks like this can stay a while open, I created a module to wrap the at CLI command in a reusable way: https://registry.terraform.io/modules/StefanSchoof/static-website/azurerm

**Comment by tombuildsstuff** [comment]
Upstream issue: https://github.com/tombuildsstuff/giovanni/issues/17

**Comment by WillAtHashicorp** [comment]
Hi there team.  Just wondering if there's any progress on this?

**Comment by revelfire** [comment]
I'm using Terraform Cloud which does not have the `az` command available so the workarounds here won't do the job. My only option is to perhaps delegate to a secondary script run somewhere else (such as Azure Devops) which frankly sounds like a total headache and creates invalid state.
Seems like this ought to be a priority @WillAtHashicorp  

I've tried for a little too long to get az cli working on TF Cloud but there's no standalone install I can pull and configure that'll work on the runner as far as I've been able to manage.

**Comment by GarethOates** [comment]
I too am eagerly awaiting progress on this issue.  What's the latest?

**Comment by timja** [comment]
This is now unblocked, I added support for it to the upstream library in https://github.com/tombuildsstuff/giovanni/pull/21
Which was released in 0.7.0

Thanks for the review @tombuildsstuff  :)

**Comment by tombuildsstuff** [comment]
@timja thanks again for that PR - I'm going to open a PR shortly to vendor v0.7.0 into this Provider, at which point it should be possible for someone to take a look at this :)

**Comment by Paul2e** [comment]
Exciting news! Can't wait to use it!

**Comment by a11smiles** [comment]
So, when will this be added to the schema `https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/storage/resource_arm_storage_account.go`?

**Comment by aqche** [comment]
Opened a PR to update the schema 👍
